src/lib/teads-curve/index.ts

```diff
@@ -0,0 +1,134 @@
+import {IImpactModelInterface} from '../interfaces';
+import Spline from 'typescript-cubic-spline';
+import {KeyValuePair} from '../../types/boavizta';
+
+export class TeadsCurveModel implements IImpactModelInterface {
+  // Defined for compatibility. Not used in TEADS.
+  authParams: object | undefined;
+  // name of the data source
+  name: string | undefined;
+  // tdp of the chip being measured
+  tdp = 100;
+  // default power curve provided by the Teads Team
+  curve: number[] = [0.12, 0.32, 0.75, 1.02];
+  // default percentage points
+  points: number[] = [0, 10, 50, 100];
+  // spline interpolation of the power curve
+  spline: Spline = new Spline(this.points, this.curve);
+
+  /**
+   * Defined for compatibility. Not used in TEADS.
+   */
+  authenticate(authParams: object): void {
+    this.authParams = authParams;
+  }
+
+  /**
+   *  Configures the TEADS Plugin for IEF
+   *  @param {string} name name of the resource
+   *  @param {Object} staticParams static parameters for the resource
+   *  @param {number} staticParams.tdp Thermal Design Power in Watts
+   */
+  async configure(
+    name: string,
+    staticParams: object | undefined = undefined
+  ): Promise<IImpactModelInterface> {
+    this.name = name;
+
+    if (staticParams === undefined) {
+      throw new Error('Required Parameters not provided');
+    }
+
+    if ('tdp' in staticParams) {
+      this.tdp = staticParams?.tdp as number;
+    } else {
+      throw new Error(
+        '`tdp` Thermal Design Power not provided. Can not compute energy.'
+      );
+    }
+
+    if ('curve' in staticParams) {
+      this.curve = staticParams?.curve as number[];
+      this.spline = new Spline(this.points, this.curve);
+    }
+
+    return this;
+  }
+
+  /**
+   * Calculate the total emissions for a list of observations
+   *
+   * Each Observation require:
+   *  @param {Object[]} observations  ISO 8601 datetime string
+   *  @param {string} observations[].datetime ISO 8601 datetime string
+   *  @param {number} observations[].duration observation duration in seconds
+   *  @param {number} observations[].cpu percentage cpu usage
+   */
+  async calculate(
+    observations: object | object[] | undefined
+  ): Promise<object> {
+    if (observations === undefined) {
+      throw new Error('Required Parameters not provided');
+    }
+
+    const results: KeyValuePair[] = [];
+    if (Array.isArray(observations)) {
+      observations.forEach((observation: KeyValuePair) => {
```

Comment:
you don't like reduce? :D

Reply:
I am trying to keep it closer to Javascript rather than Typescript.

Reply:
it's javascript-specific, and you don't have to declare results variable.

Reply:
```
if (Array.isArray(observations)) {
  return observations.reduce((acc, observation) => {
    const energy = this.calculateEnergy(observation);
    acc.push({
      ...observation,
      energy,
    })
    
    return acc
  }, [])
}

```

Reply:
this solution is more like separate component and is less coupled with parent function

Reply:
Or so `acc` becomes a copy of the `observation` but with `energy` added in?

Reply:
I feel like it is increasing cognitive complexity in a library code.

Reply:
The result object contains the energy calculation + observation as given in the issue description.

Reply:
@jawache @gnanakeethan 

```js
if (Array.isArray(observations)) {
  return observations.map((observation) => {
    observation.energy = this.calculateEnergy(observation)
    
    return observation
  })
}
```
doesn't use any array push, or spread operator.

Reply:
I am using map. Reduce is something different.